core/src/block_creation_loop.rs

```diff
@@ -347,17 +349,58 @@ fn produce_window(
     Ok(())
 }
 
+/// Clamps the block producer timestamp to ensure that the leader produces a timestamp that conforms
+/// to Alpenglow clock bounds.
+fn skew_block_producer_time_nanos(
+    parent_slot: Slot,
+    parent_time_nanos: i64,
+    working_bank_slot: Slot,
+    working_bank_time_nanos: i64,
+) -> i64 {
+    let (min_working_bank_time, max_working_bank_time) =
+        BlockComponentProcessor::alpenglow_time_bounds(
+            parent_slot,
+            parent_time_nanos,
+            working_bank_slot,
+        );
+
+    working_bank_time_nanos
+        .max(min_working_bank_time)
+        .min(max_working_bank_time)
+}
+
 /// Produces a block footer with the current timestamp and version information.
 /// The bank_hash field is left as default and will be filled in after the bank freezes.
-fn produce_block_footer(block_producer_start_time: SystemTime) -> BlockFooterV1 {
-    let block_producer_time_nanos = block_producer_start_time
+fn produce_block_footer(working_bank: &WorkingBank) -> BlockFooterV1 {
+    let mut block_producer_time_nanos = working_bank
+        .start
```

Comment:
should we be using the current time? this will report the time from the start of this slot `s` but then set the value for use in slot `s + 1`